tests/test-dladdr.c

```diff
@@ -0,0 +1,157 @@
+ * @param addrsym
+ * @param expected check against expected result
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr( void *addr, char *addrsym, ExpectedResult expected_result )
+{
+    Dl_info info;
+    int  result = dladdr( addr, &info );
+    int passed = 0;
+    if (!result)
+    {
+        passed = expected_result == NoInfo;
+        printf( "check address %p which has symbol '%s' -> %s\n", addr, addrsym, passed ? "passed" : "failed" );
+        if (verbose)
+            fprintf( stderr,"could not get symbol information for address %p\n", addr );
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+        printf( "check address %p with has symbol '%s' -> %s\n",addr, addrsym, passed ? "passed" : "failed" );
+        if (verbose)
+            print_dl_info( &info );
+    }
+    return !result;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#define HMODULE void*
+#define HANDLE void*
+#define LPCVOID void*
+#define LPCSTR char*
+#define DWORD long
+#define SIZE_T long
+
+#ifdef _MSC_VER
+#define STDCALL __stdcall
+#else
+#define STDCALL __attribute__((__stdcall__))
+#endif
+
+/* link to import thunk */
+HMODULE STDCALL GetModuleHandleA (LPCSTR lpModuleName);
+/* link to directly to iat */
+__declspec(dllimport) HMODULE STDCALL LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
+SIZE_T STDCALL VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
+#define PassWithoutSymbolOnWin PassWithoutSymbol
```

Comment:
Why not to include windows.h and have those defines automatically working independently of compiler?

Reply:
Take a look at the comments - they are there to force a different type of linking for the mentioned symbols